### PR TITLE
Use TeamCity's Parallel Testing feature

### PR DESCRIPTION
**Pull request by hegyibalint**

### CODE REVIEW

File: .teamcity/src/main/kotlin/configurations/FunctionalTest.kt [hunk 1]
```diff
@@ -27,18 +43,31 @@ class FunctionalTest(
...
 
+    val assembledExtraParameters = mutableListOf(
+        functionalTestExtraParameters(functionalTestTag, testCoverage.os, testCoverage.arch, testCoverage.testJvmVersion.major.toString(), testCoverage.vendor.name),
+        "-PflakyTests=${determineFlakyTestStrategy(stage)}",
+        extraParameters,
+        when (parallelizationMethod) {
+            is ParallelizationMethod.TestDistribution -> "-DenableTestDistribution=%enableTestDistribution% -DtestDistributionPartitionSizeInSeconds=%testDistributionPartitionSizeInSeconds%"
+            else -> ""
+        }
+    ).filter { it.isNotBlank() }.joinToString(separator = " ")
```
Comment (hegyibalint) [inline .teamcity/src/main/kotlin/configurations/FunctionalTest.kt:54]: Extracted from the `applyTestDefault` from below

File: .teamcity/src/main/kotlin/model/FunctionalTestBucketGenerator.kt [hunk 2]
```diff
@@ -82,42 +83,54 @@ data class FunctionalTestBucketWithSplitClasses(
...
 
-    fun split(expectedBucketNumber: Int, enableTestDistribution: Boolean = false): List<BuildTypeBucket> {
+    private fun splitIntoBuckets(expectedBucketNumber: Int): List<BuildTypeBucket> {
         return if (expectedBucketNumber == 1) {
             listOf(
-                SmallSubprojectBucket(
-                    listOf(subProject),
-                    enableTestDistribution
-                )
+                SmallSubprojectBucket(listOf(subProject), null)
```
Comment (hegyibalint) [inline .teamcity/src/main/kotlin/model/FunctionalTestBucketGenerator.kt:133]: Looks like a mistake, as `null` signifies no parallelization (before there was the option to enable TD).
PTAL at the new function `split` if you wonder what's happening.

**Review by blindpirate — COMMENTED** [review]
@hegyibalint This is an awesome change! I have pushed some more improvements:

1. We don't need `LargeSubprojectSplitBucket` anymore. The removal is definitely a goal of adopting TeamCity Parallel Tests because test-buckets.json can be greatly simplified.
2. I like the class `ParallelizationMethod`! I made it more kotlinish and introduced a `None` instead of nullable parameters everywhere.
3. I think we should use unified terminologies: `parallelTesting` -> `TeamCityParallelTests`, avoid mixing `numberOfBuckets` and `numberOfBatches`.
4. `FunctionalTestBucket` now has only 1 subclass, so I merged it directly.
5. There's a failing unit test `macOSBuildsSubset`, I removed it after confirming we are not using `Os.MAC.ignoredSubprojects`.

PTAL!

**Comment by hegyibalint** [comment]
@bot-gradle test RFM

**Pull request merged**